react/AddressForm.tsx

```diff
@@ -1,46 +1,162 @@
-import React from 'react'
-import { Address } from 'vtex.checkout-graphql'
-import { Input } from 'vtex.styleguide'
+import React, { useState } from 'react'
+import { Input, Dropdown, ButtonPlain } from 'vtex.styleguide'
 import { useAddressContext } from 'vtex.address-context/AddressContext'
-import { LineFragment } from './typings/countryRulesTypes'
+import {
+  LineFragment,
+  Fields,
+  OptionsField,
+  Display,
+} from './typings/countryRulesTypes.d'
+import rules from './countries/rules'
+import PlaceDetails from './PlaceDetails'
+import { FormattedMessage, defineMessages } from 'react-intl'
+import NumberOption from './components/NumberOption'
 
-interface Props {
-  address: Address
-  summary: LineFragment[][]
-}
+defineMessages({
+  country: {
+    defaultMessage: '',
+    id: 'place-components.label.country',
+  },
+  state: {
+    defaultMessage: '',
+    id: 'place-components.label.state',
+  },
+  city: {
+    defaultMessage: '',
+    id: 'place-components.label.city',
+  },
+  neighborhood: {
+    defaultMessage: '',
+    id: 'place-components.label.neighborhood',
+  },
+  ['street-road-avenue']: {
+    defaultMessage: '',
+    id: 'place-components.label.street-road-avenue',
+  },
+  ['number-option']: {
+    defaultMessage: '',
+    id: 'place-components.label.number-option',
+  },
+  complement: {
+    defaultMessage: '',
+    id: 'place-components.label.complement',
+  },
+  stateAbbreviation: {
+    defaultMessage: '',
+    id: 'place-components.label.stateAbbreviation',
+  },
+})
+
+const AddressForm: StorefrontFunctionComponent<{}> = () => {
+  const { address, setAddress } = useAddressContext()
+  const [editing, setEditing] = useState<boolean>(false)
+  const { fields, display } = rules[address.country]
+  const summary = display['extended'] as LineFragment[][]
+
+  const getSummaryFields = (summary: LineFragment[][]) => {
+    let summaryFields = new Set()
+    summary.forEach((line: LineFragment[]) => {
+      line.forEach((fragment: LineFragment) => {
+        summaryFields.add(fragment.name)
+      })
+    })
+    return summaryFields
+  }
+
+  const [ignoredFields, setIgnoredFields] = useState(
+    getSummaryFields(display['compact'])
+  )
+
+  const numberHasWithoutOption = (label: string | null) => {
+    return (
+      label && label.length >= 7 && label.substr(label.length - 7) == '-option'
+    )
+  }
 
-const AddressForm: StorefrontFunctionComponent<Props> = ({
-  address,
-  summary,
-}) => {
-  const { setAddress } = useAddressContext()
+  const hasOptions = (field: OptionsField) => {
+    return field && field.options
+  }
 
   const parseLineFragment = (fragment: LineFragment) => {
```

Comment:
Remember that in a Functional Component the `render` method is its entire body. So every time the component is rendered, the functions are defined again which can compromise your component performance.

Reply:
Great, I'll have this in mind for future components, and will see how to improve the existing components. Currently, there are some nesting of functions that use some variables from the parent function, but after talking to Marcos, I think that those should either be small, or be outside. Thanks!